Arweave Web Wallet Extension Standard (#10)
ArweaveTeam/arweave-standards: wallet-extension/README.md

# Arweave Web Wallet Extensions

This file demonstrates how an Arweave Web Wallet Extension should structure it's API so that all dApps, as well as `arweave-js` can use it properly. An implementation of this can be found at [th8ta/ArConnect](https://github.com/th8ta/ArConnect).

## Logic of the wallet

Arweave web wallet extensions are not just simple middlewares to sign transactions. They allow the user to interact with dApps in a more seamless way, similar to how Web 2 logins work.

A wallet extension should provide specific permissions for the API calls that the user has to approve. To do this, the extension should show a popup window on *connect*, to ask the user for these permissions. Ideally the wallet should also provide some sort of limitation of how much the specific dApp can spend from the user's wallet, because the popup won't show on each transaction signing.

> Preferably the wallet should notify the user in an understandable way every time a transaction signing occurs.

Reviewer: ```suggestion
Preferably the wallet should notify the user in an understandable way every time a transaction signing event occurs.
```

## The global API object

> All web wallet extensions should provide (inject) a field inside the browsers `window` object called `arweaveWallet`. This can be accessed via `window.arweaveWallet`. An example `d.ts` file of this object can be found in [injected-api.d.ts](injected-api.d.ts).

Reviewer: ```suggestion
All web wallet extensions should provide (inject) a field inside the browser's `window` object called `arweaveWallet`. This can be accessed via `window.arweaveWallet`. An example `d.ts` file of this object can be found in [injected-api.d.ts](injected-api.d.ts).
```

### Indicating the injection of the API object

When injected, the extension should fire a custom event (`arweaveWalletLoaded`) for the current tab. A sample code for this can be found below, and should be added to the end of the injected script:

```ts
dispatchEvent(new CustomEvent("arweaveWalletLoaded", { detail: {} }));
```

Developers will be able to listen to this event this way:

```ts
window.addEventListener("arweaveWalletLoaded", () => {
  /** Handle wallet extension loaded event **/
});
```

### Wallet Switch Event

> If the extension provides multi-wallet support (the user is able to add more than 1 keyfiles to the extension), the wallet extension should always fire a custom event (`walletSwitch`), with the new address associated to the active keyfile as a parameter. The extension can choose to only fire the event if the app has the permissions `ACCESS_ADDRESS` or `ACCESS_ALL_ADDRESSES`.

Reviewer: ```suggestion
If the extension provides multi-wallet support (the user is able to add more than 1 keyfile to the extension), the wallet extension should always fire a custom event (`walletSwitch`), with the new address associated to the active keyfile as a parameter. The extension can choose to only fire the event if the app has the permissions `ACCESS_ADDRESS` or `ACCESS_ALL_ADDRESSES`.
```

Reviewer: We should also reference the permissions section for context.

The event can be fired at the injected script:

```ts
dispatchEvent(new CustomEvent("walletSwitch", { detail: { address: newAddress } }));
```

Developers will be able to listen to this event this way:

```ts

> window.addEventListener("arweaveWalletLoaded", () => {

Reviewer 2: event type should be `"walletSwitch"`

  const newAddress = e.detail.address;
  /** Handle wallet extension wallet switch event **/
});
```

### Important API functions

These functions are used by `arweave-js` and they should be added to the global API object. See [njected-api.d.ts](injected-api.d.ts) for more.

#### `connect(permissions, appInfo?): Promise<void>`

> Connect to the extension with a list of permissions and an optional application info (with a name and a logo).

Reviewer: We should add types to the arguments of these functions

Author: That is in the example declaration file tho

### Other API functions

#### `disconnect(): Promise<void>`

Disconnect from the wallet extension.

#### `getAllAddresses(): Promise<string[]>`

Get all addresses added to the wallet extension.

#### `encrypt(data, options): Promise<Uint8Array>`

Encrypt a string with the user's wallet.

#### `decrypt(data, options): Promise<string>`

Decrypt a string encrypted with the user's wallet.

#### `signature(data, options): Promise<string>`

Get the signature for a data array.

### Permissions

The user needs to define at least one of these permissions when connecting, in an array of strings:

- `ACCESS_ADDRESS`: for [`getActiveAddress()`](#getactiveaddress-promisestring)

- `ACCESS_ALL_ADDRESSES`: for [`getAllAaddresses()`](#getalladdresses-promisestring)

- `SIGN_TRANSACTION`: for [`sign()`](#signaturedata-options-promisestring)

- `ENCRYPT`: for [`encrypt()`](#encryptdata-options-promiseuint8array)

- `DECRYPT`: for [`decrypt()`](#decryptdata-options-promisestring)

- `SIGNATURE`: for [`signature()`](#signaturedata-options-promisestring)

## Trust

> It is important that the community and the userbase should trust in your extension, and other extensions. To gain this trust, wallet extensions should be open source and accessible to the public. Clearly explaining how it works, how it takes fees and how it guards the users' keyfiles is also necessary.

Reviewer: ```suggestion
To gain trust and ensure transparency, wallet extensions should be open source and accessible to the public. Clearly explaining how they work, if/how they take fees, and how they guard users' keyfiles is strongly recommended.
```
